diff --git a/Makefile b/Makefile
index f747190..9f8ecb0 100644
--- a/Makefile
+++ b/Makefile
@@ -1,12 +1,12 @@
 # Makefile
 
 # Default target
-.PHONY: all up build re fclean logs elk-logs elk-up elk-down elk-status elk-clean test-labels
+.PHONY: all up build re fclean logs elk-logs elk-up elk-down elk-status elk-clean test-labels generate-ip generate-secret
 
 all: elk-up up
 
 # Build without cache and start the containers
-build: generate-secret
+build: generate-secret generate-ip
 	docker compose build --no-cache
 
 up: build
@@ -21,6 +21,28 @@ elk-down:
 elk-status:
 	@bash monitoring/scripts/master_script.sh status
 
+
+generate-ip:
+	@echo "🔍 Detecting local IP..." ; \
+	if command -v ip >/dev/null 2>&1 ; then \
+		IP=$$(ip -4 -o addr show eno2 2>/dev/null | awk '{print $$4}' | cut -d/ -f1) ; \
+	else \
+		echo "⚠️ No IP detected, using localhost" ; \
+		IP=localhost ; \
+	fi ; \
+	if [ -z "$$IP" ] ; then \
+		echo "⚠️ No IP detected, using localhost" ; \
+		IP=localhost ; \
+	else \
+		echo "✅ Detected IP: $$IP" ; \
+	fi ; \
+	echo "Writing VITE_API_URL to front-end .env..." ; \
+	echo "VITE_API_URL=https://$$IP:3010" > ./src/front-end/.env ; \
+	echo "📝 Updated front-end .env:" ; \
+	cat ./src/front-end/.env
+
+
+
 # i use this to gen new .env var for JWT token at compile time, just to have to re-log 
 # and make previous users JWT-cookies invalid on api
 generate-secret:
diff --git a/src/back-end/routes/users.js b/src/back-end/routes/users.js
index 9890525..7593f4f 100644
--- a/src/back-end/routes/users.js
+++ b/src/back-end/routes/users.js
@@ -40,7 +40,7 @@ async function userRoutes(fastify, options) // Options permet de passer des vari
     const client = new OAuth2Client(
         GOOGLE_CLIENT_ID, 
         GOOGLE_CLIENT_SECRET, 
-        'https://localhost:3010/api/auth/google/callback'
+        '${API_URL}/api/auth/google/callback'
     );
 
     fastify.get('/api/test', async (request, reply) => {
diff --git a/src/back-end/server.js b/src/back-end/server.js
index 0a1a039..908f143 100644
--- a/src/back-end/server.js
+++ b/src/back-end/server.js
@@ -111,14 +111,36 @@ fastify.decorate('p_tournament', {
 });
 
 
-// CORS (our frontend)
 fastify.register(cors, {
-  origin: 'https://localhost:5173', // ✅ must match EXACTLY
+  // On fournit UNE SEULE propriété origin : la fonction
+  origin: (origin, cb) => {
+    // Pas d'Origin (curl, mobile native, etc.) — autoriser si c'est voulu
+    if (!origin) return cb(null, true);
+
+    // Autoriser exactement le front en dev
+    if (origin === 'https://localhost:5173') {
+      return cb(null, true);
+    }
+
+    // Autoriser le réseau intranet 10.x.x.x (https://10.0.0.1:5173 par ex.)
+    try {
+      if (typeof origin === 'string' && origin.startsWith('https://10.')) {
+        return cb(null, true);
+      }
+    } catch (err) {
+      return cb(new Error('Origin validation error'));
+    }
+
+    // Par défaut : rejeter la requête CORS
+    return cb(new Error('Not allowed by CORS'));
+  },
+
   credentials: true,
   methods: ['GET', 'POST', 'PUT', 'DELETE'],
-  allowedHeaders: ['Content-Type', 'Authorization']
+  allowedHeaders: ['Content-Type', 'Authorization'],
 });
 
+
 // Cookie (Registered before i18n to allow cookie-based locale detection)
 fastify.register(cookie);
 
diff --git a/src/front-end/src/app.ts b/src/front-end/src/app.ts
index 5f72006..89c677f 100644
--- a/src/front-end/src/app.ts
+++ b/src/front-end/src/app.ts
@@ -8,6 +8,9 @@ import Header from './pages/header';
 import Leaderboard from './pages/leaderboard';
 import Friends from './pages/friends';
 import Tournament from './pages/tournament';
+export const API_URL = import.meta.env.VITE_API_URL;
+// export const API_URL = 'https://localhost:3010';
+// console.log('API_URL:', API_URL);
 
 export class App {
   private router = new Router('app');
@@ -18,7 +21,9 @@ export class App {
 
   private async check_authentication(): Promise<boolean>{
     try {
-        const res = await fetch('https://localhost:3010/api/me', { credentials: 'include' });
+        console.log('API_URL:', API_URL);
+        // console.log(API_URL);
+        const res = await fetch(`${API_URL}/api/me`, { credentials: 'include' });
         const data = await res.json();
         if (data && data.success)
           {
diff --git a/src/front-end/src/env.d.ts b/src/front-end/src/env.d.ts
new file mode 100644
index 0000000..f13ec64
--- /dev/null
+++ b/src/front-end/src/env.d.ts
@@ -0,0 +1,10 @@
+/// <reference types="vite/client" />
+
+interface ImportMetaEnv {
+  readonly VITE_API_URL: string;
+  // Add other env variables here if needed
+}
+
+interface ImportMeta {
+  readonly env: ImportMetaEnv;
+}
diff --git a/src/front-end/src/main.ts b/src/front-end/src/main.ts
index c7a2c53..e07db76 100644
--- a/src/front-end/src/main.ts
+++ b/src/front-end/src/main.ts
@@ -3,6 +3,7 @@ import App from './app.ts'; // Adjust the path as needed
 
 window.addEventListener('DOMContentLoaded', () => {
   new App(); // Initializes the router and loads the first page
+  
 });
 
 
diff --git a/src/front-end/src/pages/auth.ts b/src/front-end/src/pages/auth.ts
index 94ee7ed..d022f81 100644
--- a/src/front-end/src/pages/auth.ts
+++ b/src/front-end/src/pages/auth.ts
@@ -1,12 +1,13 @@
 // src/auth/index.ts
 import Page from '../template/page.ts';
 import { i18n } from '../i18n';
+import { API_URL } from '../app.ts';
 
 export default class AuthPage extends Page {
   async render(): Promise<HTMLElement> {
     const container = document.createElement('div');
     container.id = this.id;
-
+    console.log('API_URL:', API_URL);
     container.innerHTML = `
       <style>
 
@@ -93,13 +94,13 @@ export default class AuthPage extends Page {
     const githubLoginBtn = container.querySelector('#githubLogin') as HTMLButtonElement;
     githubLoginBtn.addEventListener('click', () => {
       // Redirige vers le backend
-      window.location.href = 'https://localhost:3010/api/auth/github/login';
+      window.location.href = `${API_URL}/api/auth/github/login`;
     });
 
     const googleLoginBtn = container.querySelector("#googleLogin") as HTMLButtonElement;
     googleLoginBtn.addEventListener('click', () => {
       // Redirige vers le backend
-      window.location.href = 'https://localhost:3010/api/auth/google/login';
+      window.location.href = `${API_URL}/api/auth/google/login`;
     });
 
 
@@ -128,7 +129,7 @@ export default class AuthPage extends Page {
       }
 
       try {
-        const endpoint = isLogin ? 'https://localhost:3010/api/login' : 'https://localhost:3010/api/register';
+        const endpoint = isLogin ? `${API_URL}/api/login` : `${API_URL}/api/register`;
         const response = await fetch(endpoint, {
           method: 'POST',
           headers: { 'Content-Type': 'application/json' },
diff --git a/src/front-end/src/pages/friends.ts b/src/front-end/src/pages/friends.ts
index b724f96..8ee7364 100644
--- a/src/front-end/src/pages/friends.ts
+++ b/src/front-end/src/pages/friends.ts
@@ -1,4 +1,5 @@
 import Page from '../template/page.ts';
+import { API_URL } from '../app.ts';
 import { i18n } from '../i18n';
 
 type FriendFromAPI = {
@@ -49,7 +50,7 @@ export default class Friends extends Page {
   async FETCH_FRIENDS(): Promise<Friend[]> {
 
 	try {
-		const res = await fetch('https://localhost:3010/api/friends', {
+		const res = await fetch(`${API_URL}/api/friends`, {
 			method: 'GET',
 			credentials: 'include',
 		});
@@ -77,7 +78,7 @@ export default class Friends extends Page {
 	async FETCH_FRIEND_REQUESTS(): Promise<Friend[]> {
 
 		try {
-				const res = await fetch('https://localhost:3010/api/friends/requests', {
+				const res = await fetch('${API_URL}/api/friends/requests', {
 				method: 'GET',
 				credentials: 'include',
 			});
@@ -98,7 +99,7 @@ export default class Friends extends Page {
 	async ACCEPT_FRIEND_REQUEST(username: string): Promise<boolean> {
 
 		try {
-			const res = await fetch(`https://localhost:3010/api/friends/requests/accept/${username}`, {
+			const res = await fetch(`${API_URL}/api/friends/requests/accept/${username}`, {
 				method: 'POST',
 				credentials: 'include',
 		});
@@ -114,7 +115,7 @@ export default class Friends extends Page {
 	async DECLINE_FRIEND_REQUEST(username: string): Promise<boolean> {
 
 		try {
-			const res = await fetch(`https://localhost:3010/api/friends/requests/decline/${username}`, {
+			const res = await fetch(`${API_URL}/api/friends/requests/decline/${username}`, {
 				method: 'POST',
 				credentials: 'include',
 		});
@@ -130,7 +131,7 @@ export default class Friends extends Page {
 	async REMOVE_FRIEND(username: string): Promise<boolean> {
 
 		try {
-			const res = await fetch(`https://localhost:3010/api/friends/${username}`, {
+			const res = await fetch(`${API_URL}/api/friends/${username}`, {
 				method: 'DELETE',
 				credentials: 'include',
 			});
diff --git a/src/front-end/src/pages/header.ts b/src/front-end/src/pages/header.ts
index c0d5cf5..f89daa2 100644
--- a/src/front-end/src/pages/header.ts
+++ b/src/front-end/src/pages/header.ts
@@ -1,5 +1,6 @@
 import Page from '../template/page.ts';
 import { i18n } from '../i18n';
+import { API_URL } from '../app.ts';
 
 export default class Header extends Page {
   async render(): Promise<HTMLElement> {
@@ -85,7 +86,7 @@ export default class Header extends Page {
     });
     container.querySelector('#logoutBtn')?.addEventListener('click', async () => {
       try {
-        const res = await fetch('https://localhost:3010/api/logout', {
+        const res = await fetch(`${API_URL}api/logout`, {
           method: 'POST',
           credentials: 'include',
         });
@@ -126,7 +127,7 @@ export default class Header extends Page {
     const _count = container.querySelector('#online-count') as HTMLElement;
     const updateOnlineCount = async () => {
       try {
-        const res = await fetch('https://localhost:3010/api/users/online', {
+        const res = await fetch(`${API_URL}/api/users/online`, {
           credentials: 'include',
         });
         const json = await res.json();
diff --git a/src/front-end/src/pages/home.ts b/src/front-end/src/pages/home.ts
index e079643..6ed4421 100644
--- a/src/front-end/src/pages/home.ts
+++ b/src/front-end/src/pages/home.ts
@@ -1,5 +1,6 @@
 import Page from '../template/page.ts';
 import Profile from './profile.ts'; // adjust path as needed
+import { API_URL } from '../app.ts';
 import { i18n } from '../i18n';
 
 type User = {
@@ -12,7 +13,7 @@ type User = {
 export default class MainPage extends Page {
   async own_user(): Promise<User | null> {
     try {
-      const res = await fetch('https://localhost:3010/api/me-info', {
+      const res = await fetch(`${API_URL}/api/me-info`, {
         credentials: 'include',
       });
       if (!res.ok) throw new Error(`HTTP error: ${res.status}`);
diff --git a/src/front-end/src/pages/leaderboard.ts b/src/front-end/src/pages/leaderboard.ts
index 15040b5..3a04177 100644
--- a/src/front-end/src/pages/leaderboard.ts
+++ b/src/front-end/src/pages/leaderboard.ts
@@ -1,3 +1,4 @@
+import { API_URL } from '../app.ts';
 import Page from '../template/page.ts';
 import { i18n } from '../i18n';
 
@@ -15,7 +16,7 @@ export default class LeaderboardPage extends Page {
 
   async FETCH_PLAYERS(): Promise<Player[]> {
     try {
-      const R = await fetch('https://localhost:3010/api/leaderboard', {
+      const R = await fetch(`${API_URL}/api/leaderboard`, {
         credentials: 'include'
       });
       if (!R.ok)
diff --git a/src/front-end/src/pages/play.ts b/src/front-end/src/pages/play.ts
index 0023885..6c4f34e 100644
--- a/src/front-end/src/pages/play.ts
+++ b/src/front-end/src/pages/play.ts
@@ -1,5 +1,6 @@
 import Page from '../template/page.ts';
 import Pong from '../component/pong.ts';
+import { API_URL } from '../app.ts';
 import { i18n } from '../i18n';
 
 export default class PlayPage extends Page {
@@ -326,7 +327,7 @@ export default class PlayPage extends Page {
     // Fetch active games helper
     const fetch_games = async () => {
         try {
-            const res = await fetch('https://localhost:3010/api/pong/active-games', {
+            const res = await fetch(`${API_URL}/api/pong/active-games`, {
                 credentials: 'include'
             });
             const data = await res.json();
@@ -372,7 +373,7 @@ export default class PlayPage extends Page {
 
     // (FETCH) stats of current serv state - initial page load
     try {
-        const ress = await fetch('https://localhost:3010/api/pong/active-games', {
+        const ress = await fetch(`${API_URL}/api/pong/active-games`, {
           credentials: 'include'
         });
         const dataa = await ress.json();
@@ -417,7 +418,7 @@ export default class PlayPage extends Page {
         const ga = document.querySelector('#game-area');
         if (ga)
           ga.innerHTML = '';
-        const res = await fetch('https://localhost:3010/api/pong/status', {
+        const res = await fetch(`${API_URL}/api/pong/status`, {
           credentials: 'include'
         });
         const data = await res.json();
@@ -449,7 +450,10 @@ export default class PlayPage extends Page {
     aiBtn.onclick = async () => {
         if(queued_up) return;
         try {
-            aiSocket = new WebSocket('wss://localhost:3010/api/pong/ai/ws'); 
+            let url = API_URL;
+            url = url.replace("https://", "wss://");
+
+            aiSocket = new WebSocket(`${url}/api/pong/ai/ws`);
             aiBtn.innerText = '🤖 Connecting to AI...';
             aiBtn.disabled = true;
             
@@ -495,7 +499,9 @@ export default class PlayPage extends Page {
       try {
           if(!socket || socket == null)
           {
-            socket = new WebSocket('wss://localhost:3010/api/pong/ws');
+            let url = API_URL;
+            url = url.replace("https://", "wss://");
+            socket = new WebSocket(`${url}/api/pong/ws`);
           }
           socket.onmessage = async (msg) => {
               const data = JSON.parse(msg.data);
@@ -571,7 +577,9 @@ export default class PlayPage extends Page {
       if (!joined_game) {
         // Start local multiplayer game via backend
         try {
-          const localSocket = new WebSocket('wss://localhost:3010/api/pong/local/ws');
+          let url = API_URL;
+          url = url.replace("https://", "wss://");
+          const localSocket = new WebSocket(`${url}/api/pong/local/ws`);
           s.innerText = '🎮 Connecting...';
           s.disabled = true;
           
diff --git a/src/front-end/src/pages/profile.ts b/src/front-end/src/pages/profile.ts
index 1c3346e..264013b 100644
--- a/src/front-end/src/pages/profile.ts
+++ b/src/front-end/src/pages/profile.ts
@@ -1,6 +1,7 @@
 import Page from '../template/page.ts';
 import { i18n } from '../i18n';
           declare const Chart: any;
+import { API_URL } from '../app.ts';
 
 type FriendshipStatus = 'friends' | 'pending' | 'none';
 
@@ -15,7 +16,7 @@ export default class UserProfilePage extends Page {
 
   async getFriendshipStatus(username: string): Promise<FriendshipStatus> {
     try {
-        const res = await fetch(`https://localhost:3010/api/friends/status/${username}`, {
+        const res = await fetch(`${API_URL}/api/friends/status/${username}`, {
             credentials: 'include',
         });
 
@@ -62,7 +63,7 @@ export default class UserProfilePage extends Page {
 
 	let currentUser: { username: string } | null = null;
 	try {
-			const meRes = await fetch('https://localhost:3010/api/me-info', {
+			const meRes = await fetch(`${API_URL}/api/me-info`, {
 			credentials: 'include'
 		});
 		if (meRes.ok) {
@@ -83,7 +84,7 @@ export default class UserProfilePage extends Page {
 	}
 	let USER_DATA: any;
     try {
-        const response = await fetch(`https://localhost:3010/${user_api_call}`, {
+        const response = await fetch(`${API_URL}/${user_api_call}`, {
           credentials: 'include'
         });
         if (!response.ok) {
@@ -402,7 +403,7 @@ export default class UserProfilePage extends Page {
         const newName = prompt("New username:", USER_DATA?.username);
         if (newName && newName !== USER_DATA?.username) {
           try {
-            const res = await fetch('https://localhost:3010/api/user', {
+            const res = await fetch(`${API_URL}/api/user`, {
               method: 'POST',
               headers: { 'Content-Type': 'application/json' },
               credentials: 'include',
@@ -466,7 +467,7 @@ export default class UserProfilePage extends Page {
               }
               
               try {
-                const res = await fetch('http://localhost:3010/api/user/avatar', {
+                const res = await fetch(`${API_URL}/api/user/avatar`, {
                   method: 'POST',
                   headers: { 'Content-Type': 'application/json' },
                   credentials: 'include',
@@ -521,7 +522,7 @@ export default class UserProfilePage extends Page {
 
 			try {
 				const route = 'api/friends/requests';
-				const response = await fetch(`https://localhost:3010/${route}`, {
+				const response = await fetch(`${API_URL}/${route}`, {
 					method: 'POST',
 					credentials: 'include',
 					headers: {
@@ -551,7 +552,7 @@ export default class UserProfilePage extends Page {
 	// Logic for activate the correct button regarding the user's relationship with the viewed profile :
 	if (!isMyProfile) {
 
-		const statusRes = await fetch(`https://localhost:3010/api/friends/status/${USER_DATA.username}`, {
+		const statusRes = await fetch(`${API_URL}/api/friends/status/${USER_DATA.username}`, {
 			credentials: 'include'
 		});
 		const statusData = await statusRes.json();
@@ -592,7 +593,7 @@ export default class UserProfilePage extends Page {
 				unfriendBtn.disabled = true;
 				unfriendBtn.textContent = i18n.t('processing');
 
-				const res = await fetch(`https://localhost:3010/api/friends/${USER_DATA.username}`, {
+				const res = await fetch(`${API_URL}/api/friends/${USER_DATA.username}`, {
 					method: 'DELETE',
 					credentials: 'include'
 				});
@@ -619,7 +620,7 @@ export default class UserProfilePage extends Page {
 				pendingBtn.disabled = true;
 				pendingBtn.textContent = i18n.t('processing');
 
-				const res = await fetch(`https://localhost:3010/api/friends/requests/${USER_DATA.username}`, {
+				const res = await fetch(`${API_URL}/api/friends/requests/${USER_DATA.username}`, {
 					method: 'DELETE',
 					credentials: 'include'
 				});
@@ -642,7 +643,7 @@ export default class UserProfilePage extends Page {
 
     // fill ts with game infos
     try {
-        const g = await fetch(`https://localhost:3010/api/${USER_DATA?.username}/games`, {
+        const g = await fetch(`${API_URL}/api/${USER_DATA?.username}/games`, {
           credentials: 'include'
         });
         let u = await(g.json());
diff --git a/vault/vault-config/vault.crt b/vault/vault-config/vault.crt
new file mode 100644
index 0000000..5c9083a
--- /dev/null
+++ b/vault/vault-config/vault.crt
@@ -0,0 +1,22 @@
+-----BEGIN CERTIFICATE-----
+MIIDnDCCAoSgAwIBAgIUUawKKG4+tFkk7Hom92s/1K0m/+kwDQYJKoZIhvcNAQEL
+BQAwZTELMAkGA1UEBhMCRlIxDjAMBgNVBAgMBVBhcmlzMQ4wDAYDVQQHDAVQYXJp
+czELMAkGA1UECgwCNDIxGTAXBgNVBAsMEGZ0X3RyYW5zY2VuZGVuY2UxDjAMBgNV
+BAMMBXZhdWx0MB4XDTI1MTEyMTE1MTcyMloXDTI2MTEyMTE1MTcyMlowZTELMAkG
+A1UEBhMCRlIxDjAMBgNVBAgMBVBhcmlzMQ4wDAYDVQQHDAVQYXJpczELMAkGA1UE
+CgwCNDIxGTAXBgNVBAsMEGZ0X3RyYW5zY2VuZGVuY2UxDjAMBgNVBAMMBXZhdWx0
+MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA3pyCnHyr6LXzoWYIHBZt
+mFHdF0UAWqqh48cqccDwaMIe+CzAVjROa8sXC3Okol1bfYaQutD7cb3enQr83BEi
+RWlQB6cGIUMzeevcooQ22MuBVWxktGhxqhCe1XR6rRTiHyQMJ0/Vubbq+2T0x4VR
+/o0o8j1exSZAimzWkvA8d6sswSwmNNT5/D8ista/6a4bvE0eaqjLFhzoS10VQmAW
+x21Es+UHwHyz0jx1Pp8uUzBN1YM8wvEqAEEvNH/GWWlam892Z2Lp4+L7Ic7n4boP
+zFZggzQYlis4cqjcJhrUsEZLR2qc8LX23HqFa9/yqusKFvdOYuhOm7s19kS4Etoj
+qQIDAQABo0QwQjAhBgNVHREEGjAYggV2YXVsdIIJbG9jYWxob3N0hwR/AAABMB0G
+A1UdDgQWBBQUIKji9dQgOOhqNlYbUaUldYkTcjANBgkqhkiG9w0BAQsFAAOCAQEA
+UhMD+TMs0BBF+w7WzBUOlWrNZFPANy4YUToX7lqK5xGuZfJOoZbVAW5TToDGAvVs
+QDmYmhqZO6oCTBFG2vB8T1/Mly3yIt4wcAZi0DLemfWJZEk85WnUVib2yFjzJbtV
+SMtuFZ3dXnfbME/nyE46LAADgAHB1+hSy4zJYZZhg+bBjd3cDJgTXQNDzG3z6a5K
+0JmH4nSFJ1Ym3txjLTja3CKvUdZHIl9qQv8xlNGDuJQtJbzUV7HBE3voPx9ydwxN
+wUX0TW/3u170aworw6Qhh7SnZxd+VqjCvwB74bwfcXdZpwlgqjMdyCZ0SHGsN4eq
+HMKQYhN/Ae8+C3y7w7bQ2g==
+-----END CERTIFICATE-----
